core/modules/replica/HealthMonitorThread.cc

```diff
@@ -0,0 +1,233 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "replica/HealthMonitorThread.h"
+
+// System headers
+#include <atomic>
+#include <map>
+
+// Qserv headers
+#include "replica/ClusterHealthJob.h"
+#include "util/BlockPost.h"
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+HealthMonitorThread::Ptr HealthMonitorThread::create(Controller::Ptr const& controller,
+                                                     ControlThread::CallbackType const& onTerminated,
+                                                     WorkerEvictCallbackType const& onWorkerEvictTimeout,
+                                                     unsigned int workerEvictTimeoutSec,
+                                                     unsigned int workerResponseTimeoutSec,
+                                                     unsigned int healthProbeIntervalSec) {
+    return Ptr(
+        new HealthMonitorThread(
+            controller,
+            onTerminated,
+            onWorkerEvictTimeout,
+            workerEvictTimeoutSec,
+            workerResponseTimeoutSec,
+            healthProbeIntervalSec
+        )
+    );
+}
+
+HealthMonitorThread::WorkerResponseDelay HealthMonitorThread::workerResponseDelay() const {
+    util::Lock lock(_mtx, "HealthMonitorThread::workerResponseDelay()");
+    return _workerServiceNoResponseSec;
+}
+
+void HealthMonitorThread::run() {
+ 
+    std::string const context = "HealthMonitorThread::run()";
+    {
+        util::Lock lock(_mtx, context);
+
+        for (auto&& worker: serviceProvider()->config()->workers()) {
+            _workerServiceNoResponseSec[worker]["qserv"] = 0;
+            _workerServiceNoResponseSec[worker]["replication"] = 0;
+        }
+    }
+
+    std::string const parentJobId;  // no parent jobs
+
+    while (not stopRequested()) {
+
+        // Probe hosts. Wait for completion or expiration of the job
+        // before analyzing its findings.
+
+        info("ClusterHealthJob");
+
+        std::atomic<size_t> numFinishedJobs{0};
+        std::vector<ClusterHealthJob::Ptr> jobs;
+        jobs.emplace_back(
+            ClusterHealthJob::create(
+                _workerResponseTimeoutSec,
+                controller(),
+                parentJobId,
+                [&numFinishedJobs](ClusterHealthJob::Ptr const& job) {
+                    ++numFinishedJobs;
+                }
+            )
+        );
+        jobs[0]->start();
+
+        track<ClusterHealthJob>("ClusterHealthJob", jobs, numFinishedJobs);
```

Comment:
Based on use of `track` it looks like `ClusterHealthJob` could have a `static std::string name() { return "ClusterHealthJob"; } that gets called inside the track Function? It doesn't make it any less lines of code but may clean up the api a little, if the job name is going to be the same as the class name.

Reply:
That'a a good suggestion. It will require some work though. Right now each subclass of class `Job` has its own unique name which is used 1) to identify the job in the log files, and 2) is stored in the database as a key identifying a type of the job. That name is passed to the base class `Job` by the subclass's constructor. Here is an example:
```
ClusterHealthJob::ClusterHealthJob(unsigned int timeoutSec,
                                   Controller::Ptr const& controller,
                                   std::string const& parentJobId,
                                   CallbackType const& onFinish,
                                   Job::Options const& options)
    :   Job(controller,
            parentJobId,
            "CLUSTER_HEALTH",
```
This name is reported by the base class's method:
```
std::string const& Job::name() const { return _name; }
```
Therefore, calling that method by method `track` could be an option. A problem here is that (as you have correctly noticed) I would need to add a static method to class `ClusterHealthJob` to return the very same name `CLUSTER_HEALTH`.

I will consider this possibility.

Reply:
Done